game/src/ltn/algorithms.rs

```diff
@@ -0,0 +1,146 @@
+    pub fn from_road(map: &Map, start: RoadID) -> Neighborhood {
+        assert!(Neighborhood::is_interior_road(start, map));
+
+        // Do a simple floodfill from this road, stopping anytime we find a major road
+        let mut interior = BTreeSet::new();
+        let mut perimeter = BTreeSet::new();
+        let mut borders = BTreeSet::new();
+
+        // We don't need a priority queue
+        let mut visited = BTreeSet::new();
+        let mut queue = vec![start];
+        interior.insert(start);
+
+        while !queue.is_empty() {
+            let current = map.get_r(queue.pop().unwrap());
+            if visited.contains(&current.id) {
+                continue;
+            }
+            visited.insert(current.id);
+            for i in [current.src_i, current.dst_i] {
+                let (minor, major): (Vec<&RoadID>, Vec<&RoadID>) = map
+                    .get_i(i)
+                    .roads
+                    .iter()
+                    .partition(|r| Neighborhood::is_interior_road(**r, map));
+                if major.is_empty() {
+                    for r in minor {
+                        interior.insert(*r);
+                        queue.push(*r);
+                    }
+                } else {
+                    borders.insert(i);
+                    perimeter.extend(major);
+                }
+            }
+        }
+
+        let mut n = Neighborhood {
+            interior,
+            perimeter,
+            borders,
+
+            modal_filters: BTreeSet::new(),
+            rat_runs: Vec::new(),
+        };
+        n.rat_runs = n.find_rat_runs(map);
+        n
+    }
+
+    pub fn toggle_modal_filter(&mut self, map: &Map, r: RoadID) {
+        if self.modal_filters.contains(&r) {
+            self.modal_filters.remove(&r);
+        } else {
+            self.modal_filters.insert(r);
+        }
+        self.rat_runs = self.find_rat_runs(map);
+    }
+
+    pub fn is_interior_road(r: RoadID, map: &Map) -> bool {
+        let road = map.get_r(r);
+        road.get_rank() == RoadRank::Local
+            && road
+                .lanes
+                .iter()
+                .any(|l| PathConstraints::Car.can_use(l, map))
+    }
+
+    // Just returns a sampling of rat runs, not necessarily all of them
+    fn find_rat_runs(&self, map: &Map) -> Vec<RatRun> {
+        // Just flood from each border and see if we can reach another border.
+        //
+        // We might be able to do this in one pass, seeding the queue with all borders. But I think
+        // the "visited" bit would get tangled up between different possibilities...
+        let mut runs: Vec<RatRun> = self
+            .borders
+            .iter()
+            .flat_map(|i| self.rat_run_from(map, *i))
+            .collect();
+        runs.sort_by(|a, b| a.length_ratio.partial_cmp(&b.length_ratio).unwrap());
+        runs
+    }
+
+    fn rat_run_from(&self, map: &Map, start: IntersectionID) -> Option<RatRun> {
```

Comment:
The algorithm to detect rat runs is straightforward graph search. We start a separate search from every entry/exit to the neighborhood. The search floodfills along the interior of the neighborhood. When it finds another entry/exit, it stitches together the path and returns it. That's it.

To produce less silly / circuitous rat-runs, we should switch to Dijkstra's and weight by distance or time to cross, not just the number of segments.
Technically a rat run, but a very silly one:
![Screenshot from 2021-09-23 13-35-27](https://user-images.githubusercontent.com/1664407/134745248-a27a77da-6067-4435-af7a-bb41dee8274d.png)

Reply:
> To produce less silly / circuitous rat-runs, we should switch to Dijkstra's and weight by distance or time to cross, not just the number of segments.

Big +1 to that. I think a simple rule that routes can only be defined as a rat run if they are shorter than the shortest trip between start and end points along major roads. By that definition it's not a rat run... There could also be other rules/definitions but sounds like those are fairly easy to implemenet.

Reply:
I think leaving room for human judgment is important. This is an extreme example. But even if it's slower to cut through the neighborhood, somebody might do it if the perimeter roads are very clogged up. Michael even suggested letting somebody manually draw or indicate a rat-run that they've observed in real life.

I think in the short-term, this tool will expose lots of possible problems and use a better score function to gauge likeliness. Then, like in the GIF in this PR, the human can judge if any of the predicted rat runs are actually problems or not. The real-world solution doesn't always have to be "block off every single entry into the neighborhood except for one."